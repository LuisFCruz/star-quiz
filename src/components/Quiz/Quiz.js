import { AppBar, Toolbar, withStyles } from '@material-ui/core';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { fetchCharacters, startTimer } from '../../actions';
import ListCharacter from '../ListCharacter/ListCharacter';
import Logo from '../../components/Logo/Logo';
import Modal from '../../components/Modal/Modal';
import ModalEndGame from '../../components/ModalEndGame/ModalEndGame';
import Pagination from '../../components/Pagination/Pagination';
import Timer from '../../components/Timer/Timer';

const styles = {
  header: {
    background: '#FFAB00',
    color: '#000'
  },
  bar: {
    display: 'flex',
    justifyContent: 'space-between',
    paddingTop: '10px',
    paddingBottom: '10px',
  },
  media: {
    width: '64px',
  }
}

class Quiz extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    characters: PropTypes.array,
    maxPage: PropTypes.number,
    finished: PropTypes.bool,
    fetchCharacters: PropTypes.func,
    startTimer: PropTypes.func,

  };

  componentDidMount() {
    this.loadQuiz();
  }

  loadQuiz = async () => {
    try {
      await this.props.fetchCharacters([1, 2, 3]);
      this.props.startTimer(true);
    }
    catch(err) {
      console.log(err);
    }
  }

  render() {
    const { classes, maxPage, finished } = this.props;
    
    return (
      <div>
        <AppBar
          position="static"
          className={classes.header}
        >
          <Toolbar className={classes.bar}>
            <Link to="/">
              <Logo
                image="/assets/images/rebel.png"
                className={classes.media}
              />
            </Link>
            <Timer duration={120} />
          </Toolbar>
        </AppBar>
        <ListCharacter />
        <Pagination min={1} max={maxPage} disabled={finished}/>
        <Modal />
        <ModalEndGame />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const { characters, finished } = state;

  const maxPage = Math.ceil(characters.length / 10);

  return { characters, maxPage, finished };
}

export default connect(mapStateToProps, { fetchCharacters, startTimer })(withStyles(styles)(Quiz));
