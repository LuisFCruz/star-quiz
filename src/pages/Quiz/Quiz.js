import { AppBar, Toolbar, withStyles } from '@material-ui/core';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { savePlayer } from '../../apis/starwars-api';
import CardCharacter from '../../components/CardCharacter/CardCharacter';
import Logo from '../../components/Logo/Logo';
import Modal from '../../components/Modal/Modal';
import ModalEndGame from '../../components/ModalEndGame/ModalEndGame';
import Pagination from '../../components/Pagination/Pagination';
import Progress from '../../components/Progress/Progress';
import Timer from '../../components/Timer/Timer';
import { fetchCharacters } from '../../actions';

const styles = {
  header: {
    background: '#FFAB00',
    color: '#000'
  },
  bar: {
    display: 'flex',
    justifyContent: 'space-between',
    paddingTop: '10px',
    paddingBottom: '10px',
  },
  media: {
    width: '64px',
  },
  clock: {
    fontWeight: '700',
    fontSize: '30px',
  },
  grid: {
    display: 'grid',
    gridGap: '10px',
    gridTemplateColumns: 'repeat(auto-fill, minmax(180px, 1fr))',
    padding: '20px',
    maxWidth: '1000px',
    margin: '0 auto',
  }
}

class Quiz extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
  };

  constructor() {
    super();
    this.state = {
      modal: {
        open: false,
        character: null
      },
      startTime: false,
      finished: false,
      total: 0,
      progress: false,
    }
    this.characters = {};
  }

  setModal(open, character) {
    const modal = { open, character };
    this.setState({ modal });
  }
  
  componentDidMount() {

    this.loadQuiz();
  }

  loadQuiz = async () => {
    this.setState({ progress: true });
    try {
      await this.props.fetchCharacters([1, 2, 3]);
      this.updatePage();
      this.updateMaxPage();
      this.setState({ startTime: true });
    }
    catch(err) { }
    finally {
      this.setState({ progress: false });
    }
  }

  updateMaxPage = () => {
    const max = Object.keys(this.characters).length / 10;
    const { pagination } = this.state;
    this.setState({ pagination: { ...pagination, max } });
  }

  handleModalClickOpen = async (id) => {
    const character = this.characters[id];
    this.characters[id] = { ...character, helped: true };
    this.setModal(true, character)
  };

  handleModalClose = () => {
    this.setModal(false, null);
  };

  handleFinish = () => {
    const finished = true;

    const answered = Object.keys(this.characters).filter(key => this.characters[key].answered);
    const total = answered
      .map(key => this.characters[key])
      .reduce((aggr, {helped}) =>  aggr + (helped ? 5 : 10), 0);

    this.setState({ total, finished, modal: { open: true } });
  }

  handleFinishClose = () => {
    this.setState({ modal: { open: false } });
  }

  handleSavePlayer = (player) => {
    const { total } = this.state
    if (player.name) {
      savePlayer({...player, total});
    }

    this.handleFinishClose();
  }

  render() {
    const time = 2;
    const { classes, characters, maxPage } = this.props

    const {
      modal,
      startTime,
      finished,
      total,
      progress,
    } = this.state;

    let modalContent = <Modal />;
    let progressContent = '';

    if (progress) {
      progressContent = <Progress />
    }

    if (modal.open && finished) {
      modalContent = <ModalEndGame
        score={total}
        {...modal}
        onClose={this.handleFinishClose}
        onConfirm={this.handleSavePlayer}
      />
    }

    return (
      <div>
        {progressContent}
        <AppBar
          position="static"
          className={classes.header}
        >
          <Toolbar className={classes.bar}>
            <Link to="/">
              <Logo
                image="/assets/images/rebel.png"
                className={classes.media}
              />
            </Link>
            <Timer
              time={time}
              start={startTime}
              className={classes.clock}
              onFinish={this.handleFinish}
            />
          </Toolbar>
        </AppBar>
        <div className={classes.grid}>
          {
            characters.map((character) =>
              <CardCharacter
                key={character.id}
                character={character}
                disabled={finished}
              />
            )
          }
          {modalContent}
        </div>
        <Pagination min={1} max={maxPage} disabled={finished}/>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const { characters: allCharacter, page } = state;

  const start = (page - 1) * 10;
  const end = (page * 10);
  
  const characters = allCharacter.slice(start, end);
  const maxPage = Math.ceil(allCharacter.length / 10);

  return { characters, maxPage };
}

export default connect(mapStateToProps, { fetchCharacters })(withStyles(styles)(Quiz));
