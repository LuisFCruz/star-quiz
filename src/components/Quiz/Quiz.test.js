import { shallow } from 'enzyme';
import React from 'react';

import { Quiz } from './Quiz';

describe('<Quiz />', () => {
  let wrapper;
  let component;

  beforeEach(() => {
    const props = {
      classes: {},
      maxPage: 3,
      finished: false,
      fetchCharacters: jest.fn(),
      startTimer: jest.fn()
    };
    wrapper = shallow(<Quiz {...props} />);
    component = wrapper.instance();
  });

  test('smoke test', () => {
    expect(Quiz).toBeDefined();
    expect(component).toBeInstanceOf(Quiz);
  });

  test('render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have call fetchCharacters() and startTimer()', () => {
    expect(component.props.fetchCharacters).toHaveBeenCalledTimes(1);
    expect(component.props.startTimer).toHaveBeenCalledTimes(1);
  });

  test('should not have call fetchCharacters() and startTimer()', () => {
    const props = {
      classes: {},
      maxPage: 3,
      finished: false,
      fetchCharacters: jest.fn().mockRejectedValue(''),
      startTimer: jest.fn()
    };
    const localWrapper = shallow(<Quiz {...props} />);
    const localComponent = localWrapper.instance();

    expect(localComponent.props.fetchCharacters).toHaveBeenCalledTimes(1);
    expect(localComponent.props.startTimer).toHaveBeenCalledTimes(0);
  });
});
