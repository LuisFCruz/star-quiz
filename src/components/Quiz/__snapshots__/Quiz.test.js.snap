// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Quiz /> render 1`] = `
<div>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Toolbar)>
      <Link
        replace={false}
        to="/"
      >
        <Logo
          image="/assets/images/rebel.png"
        />
      </Link>
      <Connect(WithStyles(Timer))
        duration={120}
      />
    </WithStyles(Toolbar)>
  </WithStyles(AppBar)>
  <Connect(WithStyles(ListCharacter)) />
  <Connect(Pagination)
    disabled={false}
    max={3}
    min={1}
  />
  <Connect(WithStyles(Modal)) />
  <Connect(WithStyles(ModalEndGame)) />
</div>
`;
